### PR TITLE
[algo] feat: add cispo

### PR DESCRIPTION
**Pull request by xvlincaigou**
### What does this PR do?

This PR adds the CISPO(https://arxiv.org/pdf/2506.13585) algorithm and a corresponding recipe to verl.

### Checklist Before Starting

- [ ✅ ] Search for similar PRs. Paste at least one query link here: ...
- [✅] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [✅] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [✅] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by wuxibin89** [comment]
Could you please provide wandb metrics for the algorithm?

**Comment by xvlincaigou** [comment]
Setup: Qwen2.5-0.5B-Base, total batch size 256, mini-batch size 128, context len 8192, dataset gsm8k.
Here task2_2_1 is GRPO and task2_2_2 is CISPO.
@wuxibin89 
image
image
image
image
image

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -1380,6 +1380,65 @@ def compute_policy_loss_geo_mean(
...
+    pg_losses = -clipped_ratio_sg * advantages * log_prob
+
+    # Track clipping statistics
+    pg_clipfrac = verl_F.masked_mean((ratio != clipped_ratio).float(), response_mask)
+
+    # Apply rollout importance sampling weights if provided
+    if rollout_is_weights is not None:
+        pg_losses = pg_losses * rollout_is_weights
+
+    pg_loss = agg_loss(loss_mat=pg_losses, loss_mask=response_mask, loss_agg_mode=loss_agg_mode)
```
Comment (wuxibin89) [inline verl/trainer/ppo/core_algos.py]: `**config.global_batch_info` should be passed into agg_loss, see other policy loss implementation.

**Comment by xvlincaigou** [comment]
@wuxibin89 I resolved your comment, thank you!

**Review by wuxibin89 — APPROVED** [review]

File: examples/cispo_trainer/run_cispo_qwen2_5_0_5b_gsm8k.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
+set -x
```
Comment (wuxibin89) [inline examples/cispo_trainer/run_cispo_qwen2_5_0_5b_gsm8k.sh:1]: Please move this script to `examples/cispo_trainer`
